File: 340/longest-substring-with-at-most-k-distinct-characters.c
//好像可以用昨天用到的方法，拉著框框前進
int lengthOfLongestSubstringKDistinct(char * s, int k){
    int N = strlen(s);
    int ascii[256] = {}; //都設為0
    int left = 0, right = 0, kd = k; //kd: 功德值
    for(int i=0; i<N; i++){
        char c = s[i];
        if(ascii[c]==0){ //有1個新的字母出現了
            ascii[c]++;
            kd--; //用掉1個功德值
            if(kd>=0){ //功德值足夠，還可以增長
                right++;
            }else{ //功德值不足，只好吐出
                char c2 = s[left];
                ascii[c2]--;
                if(ascii[c2]==0) kd++;//又恢復一個功德值
                right++;
                left++;
            }
        }else{ //沒有增加新的字母
            ascii[c]++; //加的是舊的字母
            if(kd>=0){
                right++;
            }else{
                char c2 = s[left];
                ascii[c2]--;
                if(ascii[c2]==0) kd++;
                right++;
                left++;
            }
        }
    }
    return right - left;
}

File: 340/longest-substring-with-at-most-k-distinct-characters.py
# 最多有k種不同字母的substring，最長有多長
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        counter = collections.Counter()
        counterN = 0 # 加速用的，判斷「有幾種不同的字母」
        ans = 0
        left = 0
        for right in range(len(s)):
            counter[s[right]] += 1 # 統計有幾種不同的長度
            if counter[s[right]]==1: counterN += 1 # 加速用的
            while counterN>k: # 超過，就把尾巴縮短
                counter[s[left]] -= 1
                if counter[s[left]]==0: counterN -= 1 # 加速用的
                left += 1
            # 現在就是合法的長度
            ans = max(ans, right-left+1) 
        return ans
